src/controllers/restatedeployment/reconcilers/knative.rs

```diff
@@ -0,0 +1,938 @@
+fn validate_knative_containers(containers: &[serde_json::Value]) -> Result<()> {
+    for (idx, container) in containers.iter().enumerate() {
+        // Validate port names
+        // Knative only allows port names: empty, "h2c", or "http1"
+        if let Some(ports) = container.get("ports").and_then(|p| p.as_array()) {
+            for port in ports {
+                if let Some(port_name) = port.get("name").and_then(|n| n.as_str()) {
+                    if port_name != "h2c" && port_name != "http1" {
+                        return Err(Error::InvalidRestateDeployment(
+                            format!(
+                                "Container {} has invalid port name '{}'. Knative only allows port names: 'h2c' or 'http1'. \
+                                Please update the port name in spec.template.spec.containers[{}].ports",
+                                container.get("name").and_then(|n| n.as_str()).unwrap_or("unknown"),
+                                port_name,
+                                idx
+                            )
+                        ));
+                    }
+                }
+            }
+        }
+    }
+
+    Ok(())
+}
+
+/// Ensure containers have Restate port 9080 with h2c protocol
+/// If port 9080 doesn't exist, add it. Otherwise, leave ports as-is.
+fn ensure_restate_port(containers: &mut Vec<serde_json::Value>) -> Result<()> {
+    for container in containers.iter_mut() {
+        let ports = container.get_mut("ports").and_then(|p| p.as_array_mut());
+
+        if let Some(ports_array) = ports {
+            // Check if port 9080 already exists
+            let has_restate_port = ports_array.iter().any(|port| {
+                port.get("containerPort")
+                    .and_then(|p| p.as_i64())
+                    .map(|p| p == 9080)
+                    .unwrap_or(false)
+            });
+
+            if !has_restate_port {
+                // Add Restate port with h2c protocol
+                ports_array.push(json!({
+                    "name": "h2c",
+                    "containerPort": 9080,
+                    "protocol": "TCP"
+                }));
+            }
+        } else {
+            // No ports defined, create array with Restate port
+            container.as_object_mut().unwrap().insert(
+                "ports".to_string(),
+                json!([{
+                    "name": "h2c",
+                    "containerPort": 9080,
+                    "protocol": "TCP"
+                }]),
+            );
+        }
+    }
+
+    Ok(())
+}
+
+/// Ensure containers have a readiness probe configured
+/// If no readiness probe exists, inject a TCP probe on port 9080 (Restate ingress port)
+/// with quick timing parameters suitable for fast-starting Restate SDK services.
+/// Preserves user-specified probes without modification.
+fn ensure_readiness_probe(containers: &mut Vec<serde_json::Value>) -> Result<()> {
```

Comment:
is it needed? the user can set it if they need it, but the system should work fine without it i think? unless knative mandates it?